README

= randomaze
random object generator.

== Synopsis

  require 'randomaze'

  # generating string.

  # useful basic functions.
  Randomaze::String.alpha     #=> "afefw2"
  Randomaze::String.alpha(10) #=> "afefw2rr0a"

  # you can generate randam string with regexp like expression.
  Randomaze::String.format('[a-z]{5}[0-9]{3}') #=> "feawr342"

  # generate time object.
  Randomaze::Time.between('2008-10-01'..'2009-01-01') #=> Fri Dec 19 21:14:12 +0900 2008

== Description

== Installation

=== Archive Installation

	rake install

=== Gem Installation

	gem install randomaze

== Features/Problems

== Copyright

Author::    Keiji, Yoshimi <walf443 at gmail.com>
Copyright:: Copyright (c) 2008 Keiji, Yoshimi
License:: you can redistribute it and/or modify it under the same terms as Ruby itself.



ChangeLog
== 0.0.1 / 2008-11-13

* initial release



spec/spec_helper.rb
require 'rubygems'
require 'spec'
$LOAD_PATH.unshift(File.expand_path(File.dirname(__FILE__) + '/../lib/'))
require 'randomaze'



spec/randomaze_spec.rb
require File.join(File.dirname(__FILE__), 'spec_helper')
require 'randomaze'

describe Randomaze do
end


spec/randomaze_time_spec.rb
require File.join(File.dirname(__FILE__), 'spec_helper')
require 'time'

describe Randomaze::Time do
  describe '#between' do
    it 'should generate time between time range' do
      from = Time.utc(2008, 11, 14)
      to   = Time.utc(2008, 12, 14)
      10.times do
        date = Randomaze::Time.between from..to
        ( from..to ).should include(date)
      end
    end

    it 'should generate time between string time range' do
      range = '2008-11-14'..'2008-12-14'
      from = Time.parse(range.first)
      to   = Time.parse(range.last)
      10.times do
        date = Randomaze::Time.between range
        ( from..to ).should include(date)
      end
    end
  end
end


lib/randomaze/string.rb
autoload :Set, 'set'

module Randomaze
  module String
    autoload :TemplateParser, 'randomaze/string/template_parser.tab'

    module Core
      def default_max_length
        @default_max_length ||= 1024
      end

      def random_length
        Randomaze.rand(default_max_length, 1)
      end

      def default_max_length= int
        @default_max_length = int
      end

      # Return +length+ length random string that contain array chars.
      # If +check+ is true, return nil
      def from_array array, length=random_length, &check
        check_result check do
          result = []
          case length
          when Range, Array
            tmp = length.to_a
            if ( tmp.first == 0 && tmp.last == 1 )
              length = Randomaze.rand.round
            elsif ( tmp.last.nil? )
              range = (tmp.first..default_max_length).to_a
              length = range[Randomaze.rand(range.size)]
            else
              length = tmp[Randomaze.rand(tmp.size)]
            end
          end
          length.times do
            result << array[Randomaze.rand(array.size)]
          end
          result.join
        end
      end

      #
      # Randomaze::String.from_template([['a'..'z',3], '__', [0..9, 2 ]] ) #=> 'abc__01
      #
      def from_template templates, options={}, &check
        check_result check do
          results = []
          templates.each do |item|
            results << _gen_random_element(item)
          end
          results.join
        end
      end

      def _gen_random_element item
        case item
        when Array
          set = item.first
          set = set.map {|i| i.to_a }.flatten
          lenght = item.last

          from_array(set, lenght)
        when Set
          pick = item.to_a[Randomaze.rand(item.size)]
          _gen_random_element(pick)
        else
          item
        end
      end
      private :_gen_random_element

      def format template, &check
        templates = Randomaze::String::TemplateParser.parse(template)
        from_template(templates, &check)
      end

      # Run +block+ and apply the result to +check_proc+.
      # If +check_proc+'s result is true, return nil. And else return +block+'s result.
      # It's useful for you to define your custom random string generator.
      #
      # you can skip generated value in block_lists like followings:
      #
      #   random_string = nil
      #   loop do
      #     random_string = YourMod.your_method(length) {|result| black_lists.include?(result) }
      #     break if random_string
      #   end
      #
      #   # do some thing with random_string.
      #
      def check_result check_proc=nil, &block
        result = block.call
        if check_proc && check_proc.call(result)
          nil
        else
          result
        end
      end

      def self.included klass
        klass.extend self
      end
    end

    module Utils
      def rand_base
        Randomaze::String
      end
    end

    module Basic
      ALPHA_UPCASE   = ('A'..'Z').to_a
      ALPHA_DOWNCASE = ('a'..'z').to_a
      ALPHA = ALPHA_UPCASE + ALPHA_DOWNCASE
      NUMBER = ('0'..'9').to_a
      HEX = NUMBER + ('a' .. 'f').to_a
      ALNUM = ALPHA + NUMBER

      def self.included klass
        klass.extend self
      end

      # For exmaple, "abcAbc".
      # If you'd like to get only downcase word, please use it with followings:
      #   Randomaze::String.alpha(6, ALPHA_DOWNCASE) #=> 'abcabc'
      #
      def alpha length=Randomaze::String.random_length, type=ALPHA, &check
        Randomaze::String.from_array(type, length, &check)
      end

      # For exmaple, "01234".
      def number length=Randomaze::String.random_length, &check
        Randomaze::String.from_array(NUMBER, length, &check)
      end

      # For exmaple, "0139af".
      def hex length=Randomaze::String.random_length, &check
        Randomaze::String.from_array(HEX, length, &check)
      end

      # For exmaple, "019abc".
      def alnum length=Randomaze::String.random_length, array=ALPHA, &check
        Randomaze::String.from_array(array + NUMBER, length, &check )
      end
    end

    module Inflection
      module Methods
        include Utils

        # For exmaple, "CamelCase", "Camel20Case".
        # If you'd like to get start from downcase char, please use it with followings:
        #   Randomaze::String.calme_case(6, ALPHA_DOWNCASE) #=> 'camelCase'
        #
        def camel_case length=Randomaze::String.random_length, start=rand_base::ALPHA_UPCASE, &check
          rand_base.check_result check do
            rand_base.alpha(1, start) + rand_base.alnum(length-1, rand_base::ALPHA)
          end
        end

        # For exmaple, "CONST_NAME". result is no starting from '_'.
        def const_name length=rand_base.random_length, &check
          weight = 3
          rand_base.check_result check do
            rand_base.alpha(1, rand_base::ALPHA_UPCASE) + rand_base.alpha(length-1, rand_base::ALPHA_UPCASE + ['_'] * weight)
          end
        end

        # For exmaple, "snake_case", "snake_20case". result is no starting from '_'.
        def snake_case length=rand_base.random_length, &check
          rand_base.check_result check do
            if length > 2
              weight = 3
              [
                rand_base.alpha(1, rand_base::ALPHA_DOWNCASE),
                rand_base.from_array(rand_base::ALPHA_DOWNCASE + rand_base::NUMBER + ['_'] * weight, length-2),
                rand_base.alpha(1, rand_base::ALPHA_DOWNCASE),
              ].join
            else
              rand_base.alpha(length, rand_base::ALPHA_DOWNCASE)
            end
          end
        end
      end

      include Methods
    end

    include Core
    include Basic
  end
end


lib/randomaze/string/template_parser.rb
#
# DO NOT MODIFY!!!!
# This file is automatically generated by racc 1.4.5
# from racc grammer file "lib/randomaze/string/template_parser.y".
#

require 'racc/parser'


require 'strscan'


module Randomaze

  module String

    class TemplateParser < Racc::Parser

module_eval <<'..end lib/randomaze/string/template_parser.y modeval..id0afb289b33', 'lib/randomaze/string/template_parser.y', 66

def self.parse str
  new.parse(str)
end

def parse str
  raise Racc::ParseError if str.empty?
  @tokens = self.class.tokenize str
  result = do_parse
  arr = []
  tmp = result
  while tmp[1].kind_of? Array
    arr << tmp.first
    tmp = tmp[1]
  end
  arr << tmp.first

  arr
end

def next_token
  @tokens.shift
end

def self.tokenize str
  tokens = []
  s = StringScanner.new str
  set_start_fg = false
  until s.eos? do
    if set_start_fg
      case
      when s.scan(/\]/)
        set_start_fg = false
        tokens.push([:SETS_END, ']' ])
      when s.scan(/([A-Za-z0-9])-([A-Za-z0-9])/)
        tokens.push([ :RANGE, [ s[1], s[2] ] ])
        if s[1] =~ /[a-zA-Z]/ && s[2] =~ /[a-zA-Z]/
          tokens.push([ :ALPHA, s[1] ])
          tokens.push([ :ALPHA, s[2] ])
        elsif s[1] =~ /[0-9]/ && s[2] =~ /[0-9]/
          tokens.push([ :NUMCHAR, s[1].to_i ])
          tokens.push([ :NUMCHAR, s[2].to_i ])
        else
          raise Racc::ParseError, "#{s[0]} is wrong"
        end

      when s.scan(/[0-9]+/)
        raise Racc::ParseError, "#{s[0]} is wrong" if s[0].size > 1
        tokens.push([:NUMCHAR, s[0].to_i ])
      when s.scan(/[a-zA-Z]/)
        tokens.push([:ALPHA, s[0] ])
      when s.scan(/-/)
        tokens.push(['-', s[0] ])
      else
        char = s.getch
        tokens.push [:IDENT, char]
      end
    else
      case
      when s.scan(/\[/)
        set_start_fg = true
        tokens.push([:EXPR_START, '[' ])
        tokens.push([:SETS_START, '[' ])
      when s.scan(/\{/)
        tokens.push([:COUNT_START, '{' ])
      when s.scan(/\}/)
        tokens.push([:COUNT_END, '}' ])
      when s.scan(/[0-9]+/)
        tokens.push([:NUMBER, s[0].to_i ])
      when s.scan(%r![^\[\{\}\]]+!)
        tokens.push [:IDENT_WORD, s[0]]
      else
        char = s.getch
        tokens.push [char, char]
      end
    end
  end

  tokens
end

..end lib/randomaze/string/template_parser.y modeval..id0afb289b33

##### racc 1.4.5 generates ###

racc_reduce_table = [
 0, 0, :racc_error,
 0, 14, :_reduce_none,
 2, 14, :_reduce_2,
 1, 15, :_reduce_none,
 3, 15, :_reduce_4,
 3, 16, :_reduce_5,
 3, 17, :_reduce_6,
 0, 18, :_reduce_none,
 2, 18, :_reduce_8,
 1, 19, :_reduce_none,
 1, 19, :_reduce_10,
 3, 20, :_reduce_11,
 1, 21, :_reduce_none,
 1, 21, :_reduce_none,
 1, 22, :_reduce_none,
 1, 22, :_reduce_none ]

racc_reduce_n = 16

racc_shift_n = 27

racc_action_table = [
    14,    16,     9,    11,    14,    16,     9,    11,     1,     2,
     9,    11,     9,    11,     1,     2,    20,    21,    18,     7,
    24,     5,    26 ]

racc_action_check = [
    13,    13,    13,    13,     5,     5,     5,     5,     4,     4,
    14,    14,    23,    23,     0,     0,     7,    12,     6,     3,
    18,     2,    24 ]

racc_action_pointer = [
    12,   nil,    17,    19,     6,    -5,    12,    16,   nil,   nil,
   nil,   nil,    12,    -9,    -1,   nil,   nil,   nil,    13,   nil,
   nil,   nil,   nil,     1,    14,   nil,   nil ]

racc_action_default = [
    -1,    -3,   -16,   -16,    -1,    -7,   -16,   -16,    -2,   -14,
   -13,   -15,   -16,    -7,   -16,    -9,   -12,   -10,   -16,    -4,
    27,    -5,    -8,   -16,   -16,   -11,    -6 ]

racc_goto_table = [
    23,     3,    12,     6,    19,     8,   nil,   nil,   nil,    25,
    22 ]

racc_goto_check = [
     9,     1,     5,     3,     4,     1,   nil,   nil,   nil,     9,
     5 ]

racc_goto_pointer = [
   nil,     1,   nil,     1,    -2,    -3,   nil,   nil,   nil,   -14 ]

racc_goto_default = [
   nil,   nil,     4,   nil,   nil,   nil,    13,    15,    17,    10 ]

racc_token_table = {
 false => 0,
 Object.new => 1,
 :IDENT_WORD => 2,
 :EXPR_START => 3,
 :SETS_START => 4,
 :SETS_END => 5,
 :COUNT_START => 6,
 :NUMBER => 7,
 :COUNT_END => 8,
 :RANGE => 9,
 :IDENT => 10,
 :ALPHA => 11,
 :NUMCHAR => 12 }

racc_use_result_var = true

racc_nt_base = 13

Racc_arg = [
 racc_action_table,
 racc_action_check,
 racc_action_default,
 racc_action_pointer,
 racc_goto_table,
 racc_goto_check,
 racc_goto_default,
 racc_goto_pointer,
 racc_nt_base,
 racc_reduce_table,
 racc_token_table,
 racc_shift_n,
 racc_reduce_n,
 racc_use_result_var ]

Racc_token_to_s_table = [
'$end',
'error',
'IDENT_WORD',
'EXPR_START',
'SETS_START',
'SETS_END',
'COUNT_START',
'NUMBER',
'COUNT_END',
'RANGE',
'IDENT',
'ALPHA',
'NUMCHAR',
'$start',
'exprs',
'expr',
'sets_expr',
'count_expr',
'sets',
'set',
'range',
'ident',
'alnum']

Racc_debug_parser = false

##### racc system variables end #####

 # reduce 0 omitted

 # reduce 1 omitted

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 8
  def _reduce_2( val, _values, result )
              result = val
   result
  end
.,.,

 # reduce 3 omitted

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 20
  def _reduce_4( val, _values, result )
              first = val.shift
              result = nil
              if first == '['
                result = val
              else
                result = first
              end
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 31
  def _reduce_5( val, _values, result )
              array = []
              val[1].flatten.each do |range|
                next unless range
                array << range
              end

              result = array.uniq
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 36
  def _reduce_6( val, _values, result )
              result = val[1]
   result
  end
.,.,

 # reduce 7 omitted

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 42
  def _reduce_8( val, _values, result )
              result = val
   result
  end
.,.,

 # reduce 9 omitted

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 48
  def _reduce_10( val, _values, result )
              result = val
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 53
  def _reduce_11( val, _values, result )
              result = val[1]..val[2]
   result
  end
.,.,

 # reduce 12 omitted

 # reduce 13 omitted

 # reduce 14 omitted

 # reduce 15 omitted

 def _reduce_none( val, _values, result )
  result
 end

    end   # class TemplateParser

  end   # module String

end   # module Randomaze


lib/randomaze/string/template_parser.tab.rb
#
# DO NOT MODIFY!!!!
# This file is automatically generated by racc 1.4.5
# from racc grammer file "lib/randomaze/string/template_parser.y".
#

require 'racc/parser'


require 'strscan'
require 'set'


module Randomaze

  module String

    class TemplateParser < Racc::Parser

module_eval <<'..end lib/randomaze/string/template_parser.y modeval..id128cf0f287', 'lib/randomaze/string/template_parser.y', 135

def self.parse str
  new.parse(str)
end

def parse str
  raise Racc::ParseError if str.empty?
  @tokens = self.class.tokenize str
  result = do_parse
  arr = []
  tmp = result
  while tmp[1].kind_of? Array
    arr << tmp.first
    tmp = tmp[1]
  end
  arr << tmp.first

  arr
end

def meta_set str
  case str
  when '\w'
    # FIXME: \w is not generate multi byte character.
    ['a'..'z', 'A'..'Z', 0..9, '_']
  when '\d'
    [0..9]
  when '\s'
    [' ', "\t", "\n", "\r", "\f"]
  else
    raise Racc::ParseError, "not support meta character #{first}"
  end
end

def next_token
  @tokens.shift
end

def self.tokenize str
  tokens = []
  s = StringScanner.new str
  state = :default
  until s.eos? do
    case state
    when :default
      case
      when s.scan(/\\[a-z]/)
        tokens.push([:META_SET, s[0]])
      when s.scan(/\./)
        tokens.push([:DOT, s[0]])
      when s.scan(/\?/)
        tokens.push([:QUESTION, s[0]])
      when s.scan(/\*(\?)?/)
        tokens.push([:ASTERISK, s[0]])
      when s.scan(/\+(\?)?/)
        tokens.push([:PLUS, s[0]])
      when s.scan(/\[/)
        state = :set_expr
        tokens.push([:EXPR_START, '[' ])
        tokens.push([:SETS_START, '[' ])
      when s.skip(/\(\?#.+\)/) # COMMENT
      when s.scan(/\((\?:)?/)
        tokens.push([:SELECT_START, '(' ])
      when s.scan(/\)/)
        tokens.push([:SELECT_END, ')' ])
      when s.scan(/\|/)
        tokens.push(['|', '|' ])
      when s.scan(/\{/)
        state = :count_expr
        tokens.push([:COUNT_START, '{' ])
      when s.scan(%r![^\)|\\\[\{\}\]]+!)
        tokens.push [:IDENT_WORD, s[0]]
      when s.scan(%r!\\([\?\[\]\{\}\*\+\^\$\\\.\|\(\)])!)
        tokens.push [:IDENT_WORD, s[1]]
      else
        char = s.getch
        tokens.push [char, char]
      end
    when :count_expr
      case
      when s.scan(/\}/)
        state = :default
        tokens.push([:COUNT_END, '}' ])
      when s.scan(/[0-9]+/)
        tokens.push([:NUMBER, s[0].to_i ])
      when s.scan(/,/)
        tokens.push([',', ','])
      else
        char = s.getch
        raise Racc::ParseError, "invalid char #{char}"
      end

    when :set_expr
      case
      when s.scan(/\\[a-z]/)
        tokens.push([:META_SET, s[0]])
      when s.scan(/\]/)
        state = :default
        tokens.push([:SETS_END, ']' ])
      when s.scan(/([\w])-([\w])/)
        tokens.push([ :RANGE, [ s[1], s[2] ] ])

      when s.scan(/[0-9]+/)
        raise Racc::ParseError, "#{s[0]} is wrong" if s[0].size > 1
        tokens.push([:NUMCHAR, s[0].to_i ])
      when s.scan(/[a-zA-Z]/)
        tokens.push([:ALPHA, s[0] ])
      when s.scan(/-/)
        tokens.push(['-', s[0] ])
      else
        char = s.getch
        tokens.push [:IDENT, char]
      end
    else
      raise "must not happend!!"
    end
  end

  tokens
end

..end lib/randomaze/string/template_parser.y modeval..id128cf0f287

##### racc 1.4.5 generates ###

racc_reduce_table = [
 0, 0, :racc_error,
 0, 24, :_reduce_none,
 2, 24, :_reduce_2,
 1, 25, :_reduce_none,
 1, 25, :_reduce_4,
 3, 25, :_reduce_5,
 4, 25, :_reduce_6,
 3, 25, :_reduce_7,
 1, 25, :_reduce_8,
 2, 25, :_reduce_9,
 2, 25, :_reduce_10,
 2, 25, :_reduce_11,
 3, 25, :_reduce_12,
 3, 27, :_reduce_13,
 3, 26, :_reduce_14,
 4, 26, :_reduce_15,
 5, 26, :_reduce_16,
 1, 26, :_reduce_17,
 1, 26, :_reduce_18,
 1, 26, :_reduce_19,
 0, 28, :_reduce_none,
 2, 28, :_reduce_21,
 1, 29, :_reduce_none,
 1, 29, :_reduce_23,
 1, 30, :_reduce_24,
 1, 31, :_reduce_none,
 1, 31, :_reduce_none,
 2, 31, :_reduce_none,
 1, 31, :_reduce_28,
 1, 32, :_reduce_none,
 1, 32, :_reduce_none ]

racc_reduce_n = 31

racc_shift_n = 45

racc_action_table = [
    27,     1,     3,     5,     9,    27,     6,     7,    10,    12,
    13,    28,    30,    32,    34,    25,    28,    30,    32,    34,
    25,     1,     3,     5,    21,    15,     6,     7,     1,     3,
     5,     9,    15,     6,     7,    10,    12,    13,     1,     3,
     5,     9,    20,     6,     7,    10,    12,    13,     9,    42,
    43,    39,    10,    12,    13,    36,    37,    23,    15,    19,
    41,     8,    44 ]

racc_action_check = [
    29,     0,     0,     0,     6,    19,     0,     0,     6,     6,
     6,    29,    29,    29,    29,    29,    19,    19,    19,    19,
    19,     4,     4,     4,     9,     4,     4,     4,     5,     5,
     5,     3,    22,     5,     5,     3,     3,     3,    15,    15,
    15,    18,     8,    15,    15,    18,    18,    18,    23,    37,
    37,    26,    23,    23,    23,    21,    21,    16,    16,     7,
    32,     2,    42 ]

racc_action_pointer = [
    -1,   nil,    61,    20,    19,    26,    -7,    50,    42,    12,
   nil,   nil,   nil,   nil,   nil,    36,    52,   nil,    30,    -2,
   nil,    42,    26,    37,   nil,   nil,    41,   nil,   nil,    -7,
   nil,   nil,    40,   nil,   nil,   nil,   nil,    37,   nil,   nil,
   nil,   nil,    49,   nil,   nil ]

racc_action_default = [
    -1,    -3,   -31,    -4,    -1,   -31,    -8,   -31,   -31,   -31,
   -17,   -10,   -18,   -19,    -2,   -31,   -31,   -11,    -9,   -20,
    45,   -31,    -7,    -5,   -12,   -30,   -31,   -28,   -24,   -20,
   -25,   -22,   -31,   -23,   -29,   -26,   -14,   -31,    -6,   -13,
   -21,   -27,   -31,   -15,   -16 ]

racc_goto_table = [
    11,    16,    26,    17,     2,    18,   nil,   nil,    14,   nil,
   nil,    22,    40,   nil,   nil,    24,   nil,   nil,   nil,   nil,
    38 ]

racc_goto_check = [
     3,     2,     5,     3,     1,     4,   nil,   nil,     1,   nil,
   nil,     2,     5,   nil,   nil,     3,   nil,   nil,   nil,   nil,
     3 ]

racc_goto_pointer = [
   nil,     4,    -4,    -3,    -2,   -17,   nil,   nil,   nil,   nil ]

racc_goto_default = [
   nil,   nil,     4,   nil,   nil,   nil,    29,    31,    33,    35 ]

racc_token_table = {
 false => 0,
 Object.new => 1,
 :IDENT_WORD => 2,
 :DOT => 3,
 :SELECT_START => 4,
 :SELECT_END => 5,
 "|" => 6,
 :META_SET => 7,
 :EXPR_START => 8,
 :SETS_START => 9,
 :SETS_END => 10,
 :COUNT_START => 11,
 :NUMBER => 12,
 :COUNT_END => 13,
 "," => 14,
 :QUESTION => 15,
 :ASTERISK => 16,
 :PLUS => 17,
 :RANGE => 18,
 :IDENT => 19,
 :meta_set => 20,
 :ALPHA => 21,
 :NUMCHAR => 22 }

racc_use_result_var = true

racc_nt_base = 23

Racc_arg = [
 racc_action_table,
 racc_action_check,
 racc_action_default,
 racc_action_pointer,
 racc_goto_table,
 racc_goto_check,
 racc_goto_default,
 racc_goto_pointer,
 racc_nt_base,
 racc_reduce_table,
 racc_token_table,
 racc_shift_n,
 racc_reduce_n,
 racc_use_result_var ]

Racc_token_to_s_table = [
'$end',
'error',
'IDENT_WORD',
'DOT',
'SELECT_START',
'SELECT_END',
'"|"',
'META_SET',
'EXPR_START',
'SETS_START',
'SETS_END',
'COUNT_START',
'NUMBER',
'COUNT_END',
'","',
'QUESTION',
'ASTERISK',
'PLUS',
'RANGE',
'IDENT',
'meta_set',
'ALPHA',
'NUMCHAR',
'$start',
'exprs',
'expr',
'count_expr',
'sets_expr',
'sets',
'set',
'range',
'ident',
'alnum']

Racc_debug_parser = false

##### racc system variables end #####

 # reduce 0 omitted

 # reduce 1 omitted

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 8
  def _reduce_2( val, _values, result )
              result = val
   result
  end
.,.,

 # reduce 3 omitted

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 15
  def _reduce_4( val, _values, result )
              token = val.shift
              result = [ ['a'..'z', 'A'..'Z', 0..9, '_'], 1 ]
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 19
  def _reduce_5( val, _values, result )
              result = val[1]
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 23
  def _reduce_6( val, _values, result )
              result = [ val[1], val[3] ]
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 27
  def _reduce_7( val, _values, result )
              result = Set.new([val[0], val[2]])
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 32
  def _reduce_8( val, _values, result )
              token = val.shift
              result = [ meta_set(token), 1]
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 36
  def _reduce_9( val, _values, result )
              result = [ val[1], 1]
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 41
  def _reduce_10( val, _values, result )
              token, count = val
              result = [ ['a'..'z', 'A'..'Z', 0..9, '_'], count ]
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 46
  def _reduce_11( val, _values, result )
              token, count = val
              result = [ meta_set(token), count ]
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 56
  def _reduce_12( val, _values, result )
              first = val.shift
              result = nil
              if first == '['
                result = val
              else
                result = first
              end
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 67
  def _reduce_13( val, _values, result )
              array = []
              val[1].flatten.each do |range|
                next unless range
                array << range
              end

              result = array.uniq
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 72
  def _reduce_14( val, _values, result )
               result = val[1]
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 76
  def _reduce_15( val, _values, result )
               result = [ val[1], nil ]
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 80
  def _reduce_16( val, _values, result )
               result = val[1]..val[3]
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 84
  def _reduce_17( val, _values, result )
               result = 0..1
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 88
  def _reduce_18( val, _values, result )
               result = [0, nil]
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 92
  def _reduce_19( val, _values, result )
               result = [1, nil]
   result
  end
.,.,

 # reduce 20 omitted

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 98
  def _reduce_21( val, _values, result )
              result = val
   result
  end
.,.,

 # reduce 22 omitted

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 104
  def _reduce_23( val, _values, result )
              result = val
   result
  end
.,.,

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 114
  def _reduce_24( val, _values, result )
              first,last = val.first.to_a
              if first =~ /\d/ && last =~ /\d/
                first = first.to_i
                last  = last.to_i
              end
              result = first..last
   result
  end
.,.,

 # reduce 25 omitted

 # reduce 26 omitted

 # reduce 27 omitted

module_eval <<'.,.,', 'lib/randomaze/string/template_parser.y', 123
  def _reduce_28( val, _values, result )
              first = val.shift
              result = meta_set(first)
   result
  end
.,.,

 # reduce 29 omitted

 # reduce 30 omitted

 def _reduce_none( val, _values, result )
  result
 end

    end   # class TemplateParser

  end   # module String

end   # module Randomaze
